### PR TITLE
Add Torque boundary condition

### LINKED ISSUE #17176: BendingMoment (or torque) boundary condition
## Reason
We can apply pressures and linear forces, but not bending moments or torques with respect to a reference point.

## Design
Create an integrated BC based on a Neumann BC and an action to set those up for each displacement component.

## Impact
Added capability

### PR DESCRIPTION
**Pull request by dschwen**
Closes #17176

### CODE REVIEW

File: modules/tensor_mechanics/src/bcs/TwistingMomentBC.C [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+template <bool is_ad>
+InputParameters
+TwistingMomentBCTempl<is_ad>::validParams()
+{
+  InputParameters params = TwistingMomentBCParent<is_ad>::validParams();
+  params.addClassDescription("Apply a moment around a pivot point. This should operate on "
+                             "the displaced mesh for large deformations.");
+  MooseEnum component("x=0 y=1 z=2");
+  params.addRequiredParam<Point>("origin", "Pivot point for the bending moment");
+  params.addRequiredParam<RealVectorValue>("direction", "Torque vector");
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/bcs/TwistingMomentBC.C:25]: I initially found the "direction" and "factor" parameters kind of confusing, although it made sense once I thought about it a bit. I think I'd call them "direction" and "torque", and make them both required. I

I'd use these parameter names and descriptions:
"direction": "Direction of the torque vector. The supplied vector is normalized if not a unit vector.""
"torque": "Time-dependent function defining the magnitude of the torque"

Also, normalize the user-supplied torque vector in the constructor.

File: modules/tensor_mechanics/src/postprocessors/PolarMomentOfInertia.C [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+registerMooseObject("TensorMechanicsApp", PolarMomentOfInertia);
+
+InputParameters
+PolarMomentOfInertia::validParams()
+{
+  InputParameters params = SideIntegralPostprocessor::validParams();
+  params.addRequiredParam<Point>("origin", "Axis origin");
+  params.addRequiredParam<RealVectorValue>("direction", "Axis direction");
+  params.addClassDescription("Compute the a component of the polar moment of inertia of a sideset "
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/postprocessors/PolarMomentOfInertia.C:20]: the a -> the

File: modules/tensor_mechanics/src/postprocessors/TorqueReaction.C [hunk 3]
```diff
@@ -21,7 +21,7 @@ TorqueReaction::validParams()
   InputParameters params = NodalPostprocessor::validParams();
   params.addClassDescription("TorqueReaction calculates the torque in 2D and 3D"
                              "about a user-specified axis of rotation centered"
-                             "at a user-specied origin.");
+                             "at a user-specified origin.");
   params.addRequiredParam<std::vector<AuxVariableName>>("reaction_force_variables",
                                                         "The reaction variables");
   params.addParam<RealVectorValue>(
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/postprocessors/TorqueReaction.C:27]: Maybe we should make the parameter names in this class consistent with the new ones. I don't think it's widely used -- we just made it for a regression test.

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/doc/content/source/bcs/TwistingMomentBC.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# TwistingMomentBC
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/tensor_mechanics/doc/content/source/bcs/TwistingMomentBC.md:1]: https://eng.libretexts.org/Bookshelves/Materials_Science/TLP_Library_I/07%3A_Bending_and_Torsion_of_Beams/7.06%3A_Section_6-#:~:text=Torsion%20is%20the%20twisting%20of,many%20other%20structures%20and%20components.

I'd argue that twisting the the motion itself, while twisting moment is the cause of it.

File: modules/tensor_mechanics/src/bcs/TwistingMomentBC.C [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    this->paramError("variables",
+                     "The kernel variable needs to be one of the 'displacements' variables");
+}
+
+template <>
+ADReal
+TwistingMomentBCTempl<true>::computeQpResidual()
+{
+  // local lever (distance to the origin)
+  auto local_lever = this->_ad_q_points[_qp] - _origin;
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/tensor_mechanics/src/bcs/TwistingMomentBC.C:66]: yeah

File: modules/tensor_mechanics/src/bcs/TwistingMomentBC.C (hunk 1 above)
Comment (dschwen) [reply modules/tensor_mechanics/src/bcs/TwistingMomentBC.C:25]: Hm, that would make it a lot more difficult to apply a specified torque vector. The user would have to compute the magnitude and specify that again as the `torque` parameter

File: modules/tensor_mechanics/src/auxkernels/DeflectionAngle.C [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    dr1 -= _direction * (_direction * dr1);
+    dr2 -= _direction * (_direction * dr2);
+  }
+
+  // product of the lengths
+  auto norms = std::sqrt(dr1.norm_sq() * dr2.norm_sq());
+
+  // angle between dr1 and dr2
+  if (norms > libMesh::TOLERANCE)
+    return std::acos((dr1 * dr2) / norms);
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/tensor_mechanics/src/auxkernels/DeflectionAngle.C]: Implemented and tested

**Comment by dschwen** [comment]
here still is a sign error in the application of the torque. It does not conform to the right hand rule. I'll fix that.

File: modules/tensor_mechanics/include/postprocessors/PolarMomentOfInertia.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "SideIntegralPostprocessor.h"
+
+/**
+ * Compute the a component of the second moment of the area of a sideset w.r.t. a point
```
[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/tensor_mechanics/include/postprocessors/PolarMomentOfInertia.h:15]: This still needs fixing

File: modules/tensor_mechanics/include/postprocessors/PolarMomentOfInertia.h [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+{
+public:
+  static InputParameters validParams();
+
+  PolarMomentOfInertia(const InputParameters & parameters);
+
+protected:
+  Real computeQpIntegral() override;
+
+  const Point _origin;
```
[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/tensor_mechanics/include/postprocessors/PolarMomentOfInertia.h:27]: Still needs to be done

File: modules/tensor_mechanics/src/bcs/Torque.C [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+template <bool is_ad>
+InputParameters
+TorqueTempl<is_ad>::validParams()
+{
+  InputParameters params = TorqueParent<is_ad>::validParams();
+  params.addClassDescription(
+      "Apply a moment as tractions distributed over a surface around a "
+      "pivot point. This should operate on the displaced mesh for large deformations.");
+  MooseEnum component("x=0 y=1 z=2");
+  params.addRequiredParam<Point>("origin", "Pivot point for the twisting moment");
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/bcs/Torque.C:25]: Do we need to allow for the origin to move for finite deformations? I guess there are relevant cases where it doesn't move, but it seems like that might be important for some cases.

File: modules/tensor_mechanics/src/postprocessors/PolarMomentOfInertia.C (hunk 2 above)
Comment (bwspenc) [reply modules/tensor_mechanics/src/postprocessors/PolarMomentOfInertia.C:20]: still needs to be addressed

File: modules/tensor_mechanics/src/postprocessors/TorqueReaction.C (hunk 3 above)
Comment (bwspenc) [reply modules/tensor_mechanics/src/postprocessors/TorqueReaction.C:27]: I think I prefer 'origin' and 'direction' like you have in the new code. At least they're both specifying the same quantities, although with different names now. We could go back later and update this.

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/src/bcs/Torque.C (hunk 9 above)
Comment (dschwen) [reply modules/tensor_mechanics/src/bcs/Torque.C]: The origin should be the `Point` equivalent of a Postprocessor (using the new Reporter system).  But I would wait implementing it that way until users can specify value literals in place of reporter names.

**Review by bwspenc — APPROVED** [review]

**Pull request merged**